### PR TITLE
(BIDS-2368) optimize withdrawal queries

### PR DESCRIPTION
**Pull request by LuccaBitfly**
### <samp>🤖 Generated by Copilot at a5cc42b</samp>

This pull request optimizes and enhances the withdrawal data for validators in the eth2-beaconchain-explorer app. It removes or modifies some obsolete or inefficient functions in the `db` package, and adds new functions in the `handlers` package that query different tables and subqueries. It also updates the API and the dashboard to use the new functions and display more accurate and relevant information to the users.

### CODE REVIEW

**Comment by LuccaBitfly** [comment]
I may have missed queries that could be optimized, pls check. Keep in mind that optimizing with the stats table only works if the query concerns aggregated values dependent on validators as input parameter.

**Review by Eisei24 — CHANGES_REQUESTED** [review]
Please have a look at my comments.

File: handlers/api.go [hunk 1]
```diff
@@ -1357,18 +1357,10 @@ func getGeneralValidatorInfoForAppDashboard(queryIndices []uint64) ([]interface{
 		COALESCE(validator_performance.cl_performance_365d, 0) AS performance365d,
 		COALESCE(validator_performance.cl_performance_total, 0) AS performanceTotal,
 		COALESCE(validator_performance.rank7d, 0) AS rank7d,
-		((validator_performance.rank7d::float * 100) / COALESCE((SELECT total_count FROM maxValidatorIndex), 1)) as rankpercentage,
-		w.total as total_withdrawals
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply handlers/api.go:1361]: Yes, I have discussed this with @manuelsc, the app does not use the info so I removed it

File: handlers/api.go [hunk 2]
```diff
@@ -1554,31 +1546,56 @@ func apiValidator(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	lastExportedDay, err := services.LatestExportedStatisticDay()
+	if err != nil {
+		sendServerErrorResponse(w, r.URL.String(), "error retrieving data, please try again later")
+		return
+	}
+	firstSlotToday := ((lastExportedDay+1)*utils.EpochsPerDay() - 1) * utils.Config.Chain.Config.SlotsPerEpoch // -1 because validator_stats does not include last epoch of the day for some reason
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply handlers/api.go]: fixed

File: handlers/api.go [hunk 3]
```diff
@@ -1554,31 +1546,56 @@ func apiValidator(w http.ResponseWriter, r *http.Request) {
...
+			FROM validator_stats vs
+			WHERE vs.validatorindex = ANY($1)
+			GROUP BY vs.validatorindex
+		),
+		withdrawals_summary AS (
+			SELECT
+				COALESCE(t.validatorindex, s.validatorindex) as validatorindex,
+				COALESCE(t.amount_today, 0) + COALESCE(s.total_amount, 0) as total
+			FROM today t
+			FULL OUTER JOIN stats s ON t.validatorindex = s.validatorindex
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply handlers/api.go]: fixed

File: handlers/common.go [hunk 4]
```diff
@@ -707,3 +703,36 @@ func GetWithdrawableCountFromCursor(epoch uint64, validatorindex uint64, cursor
...
 }
+
+func GetTotalWithdrawalsCount(validators []uint64) (uint64, error) {
+	var count uint64
+	validatorFilter := pq.Array(validators)
+	lastExportedDay, err := services.LatestExportedStatisticDay()
+	if err != nil {
+		return 0, fmt.Errorf("error getting latest exported statistic day for withdrawals count: %w", err)
+	}
+	firstSlotToday := ((lastExportedDay+1)*utils.EpochsPerDay() - 1) * utils.Config.Chain.Config.SlotsPerEpoch // -1 because validator_stats does not include last epoch of the day for some reason
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply handlers/common.go]: fixed

File: handlers/common.go [hunk 5]
```diff
@@ -707,3 +703,36 @@ func GetWithdrawableCountFromCursor(epoch uint64, validatorindex uint64, cursor
...
+
+	err = db.ReaderDb.Get(&count, `
+		WITH today AS (
+			SELECT count(*) as count_today
+			FROM blocks_withdrawals w
+			INNER JOIN blocks b ON b.blockroot = w.block_root AND b.status = '1'
+			WHERE w.validatorindex = ANY($1) AND w.block_slot >= $2
+		),
+		stats AS (
+			SELECT COALESCE(SUM(withdrawals)) as total_count
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply handlers/common.go]: fixed

**Review by Eisei24 — COMMENTED** [review]
`GetTotalAmountWithdrawn` is a function that could use validator_stats or does it become slower since we are not searching for specific validators?

On a side note it is also weird that it gives names to the return values but then doesn't use them.

**Comment by LuccaBitfly** [comment]
Regarding `GetTotalAmountWithdrawn`: The performance gets worse when incorporating the whole stats table, but limiting to a single day with `total` columns should improve it. So not doing it in this PR

File: handlers/common.go [hunk 6]
```diff
@@ -785,3 +781,36 @@ func getExecutionChartData(indices []uint64, currency string, lowerBoundDay uint
...
 }
+
+func GetTotalWithdrawalsCount(validators []uint64) (uint64, error) {
+	var count uint64
+	validatorFilter := pq.Array(validators)
+	lastExportedDay, err := services.LatestExportedStatisticDay()
+	if err != nil {
+		return 0, fmt.Errorf("error getting latest exported statistic day for withdrawals count: %w", err)
+	}
+	cutoffSlot := (((lastExportedDay+1)*utils.EpochsPerDay() - 1) * utils.Config.Chain.Config.SlotsPerEpoch) + 1
```
Comment (peterbitfly) [inline handlers/common.go:792]: better to use `utils.GetFirstAndLastEpochForDay(day)` for that

File: handlers/common.go [hunk 7]
```diff
@@ -785,3 +781,36 @@ func getExecutionChartData(indices []uint64, currency string, lowerBoundDay uint
 
 	return chartData, nil
 }
+
+func GetTotalWithdrawalsCount(validators []uint64) (uint64, error) {
```
Comment (peterbitfly) [inline handlers/common.go:785]: this method should be moved to the db package

**Review by peterbitfly — CHANGES_REQUESTED** [review]

File: handlers/common.go (hunk 6 above)
Comment (LuccaBitfly) [reply handlers/common.go]: fixed

File: handlers/common.go (hunk 7 above)
Comment (LuccaBitfly) [reply handlers/common.go]: fixed

**Review by peterbitfly — APPROVED** [review]

**Review by Eisei24 — APPROVED** [review]

**Pull request merged**